File: src/voyage_geo/stages/analysis/analyzers/sentiment.py
"""Sentiment analyzer — VADER-based sentiment analysis of brand mentions."""

from __future__ import annotations

import statistics

from vaderSentiment.vaderSentiment import SentimentIntensityAnalyzer

from voyage_geo.types.analysis import SentimentExcerpt, SentimentScore
from voyage_geo.types.brand import BrandProfile
from voyage_geo.types.result import QueryResult
from voyage_geo.utils.text import contains_brand, extract_sentences

vader = SentimentIntensityAnalyzer()


def _label(score: float) -> str:
    if score >= 0.05:
        return "positive"
    if score <= -0.05:
        return "negative"
    return "neutral"


class SentimentAnalyzer:
    name = "sentiment"

    def analyze(self, results: list[QueryResult], profile: BrandProfile) -> SentimentScore:
        valid = [r for r in results if not r.error and r.response]

        scored: list[dict] = []  # {text, score, provider}

        for r in valid:
            sentences = [s for s in extract_sentences(r.response) if contains_brand(s, profile.name)]
            for sentence in sentences:
                vs = vader.polarity_scores(sentence)
                scored.append({"text": sentence, "score": vs["compound"], "provider": r.provider})

        if not scored:
            return SentimentScore()

        scores = [s["score"] for s in scored]
        overall = statistics.mean(scores)
        label = _label(overall)

        stddev = statistics.stdev(scores) if len(scores) > 1 else 0.0
        sample_factor = min(len(scored) / 10, 1.0)
        variance_factor = max(0.0, 1.0 - stddev)
        confidence = round(sample_factor * variance_factor, 2)

        pos = sum(1 for s in scores if s >= 0.05)
        neg = sum(1 for s in scores if s <= -0.05)
        neu = len(scores) - pos - neg

        # By provider
        by_provider: dict[str, float] = {}
        by_provider_label: dict[str, str] = {}
        prov_groups: dict[str, list[float]] = {}
        for s in scored:
            prov_groups.setdefault(s["provider"], []).append(s["score"])
        for prov, prov_scores in prov_groups.items():
            avg = statistics.mean(prov_scores)
            by_provider[prov] = round(avg, 4)
            by_provider_label[prov] = _label(avg)

        # By category (from query ID prefix)
        by_category: dict[str, float] = {}
        cat_scores: dict[str, list[float]] = {}
        for r in valid:
            prefix = r.query_id.split("-")[0]
            cat_map = {"kw": "keyword", "ps": "persona", "cp": "competitor", "in": "intent"}
            cat = cat_map.get(prefix, "unknown")
            sentences = [s for s in extract_sentences(r.response) if contains_brand(s, profile.name)]
            for sentence in sentences:
                vs = vader.polarity_scores(sentence)
                cat_scores.setdefault(cat, []).append(vs["compound"])
        for cat, cs in cat_scores.items():
            by_category[cat] = round(statistics.mean(cs), 4)

        # Top excerpts
        sorted_scored = sorted(scored, key=lambda s: s["score"], reverse=True)
        top_positive = [
            SentimentExcerpt(text=s["text"][:200], score=s["score"], provider=s["provider"])
            for s in sorted_scored if s["score"] >= 0.05
        ][:5]
        top_negative = [
            SentimentExcerpt(text=s["text"][:200], score=s["score"], provider=s["provider"])
            for s in reversed(sorted_scored) if s["score"] <= -0.05
        ][:5]

        return SentimentScore(
            overall=round(overall, 4),
            label=label,  # type: ignore[arg-type]
            confidence=confidence,
            by_provider=by_provider,
            by_provider_label=by_provider_label,  # type: ignore[arg-type]
            by_category=by_category,
            positive_count=pos,
            neutral_count=neu,
            negative_count=neg,
            total_sentences=len(scored),
            top_positive=top_positive,
            top_negative=top_negative,
        )

File: src/stages/analysis/analyzers/sentiment.ts
import Sentiment from 'sentiment';
import type { Analyzer } from './types.js';
import type { QueryResult } from '../../../types/result.js';
import type { BrandProfile } from '../../../types/brand.js';
import type { SentimentScore, SentimentExcerpt } from '../../../types/analysis.js';
import { containsBrand, extractSentences } from '../../../utils/text.js';
import { groupBy, mean, standardDeviation } from '../statistics.js';

const sentiment = new Sentiment();

interface ScoredSentence {
  text: string;
  score: number;
  provider: string;
  category?: string;
}

function scoreLabel(score: number): 'positive' | 'neutral' | 'negative' {
  if (score > 0.05) return 'positive';
  if (score < -0.05) return 'negative';
  return 'neutral';
}

export class SentimentAnalyzer implements Analyzer {
  name = 'sentiment';

  analyze(results: QueryResult[], profile: BrandProfile): SentimentScore {
    const validResults = results.filter((r) => !r.error && r.response);
    const allSentences: ScoredSentence[] = [];

    for (const result of validResults) {
      const sentences = extractSentences(result.response)
        .filter((s) => containsBrand(s, profile.name));

      for (const sentence of sentences) {
        const analysis = sentiment.analyze(sentence);
        allSentences.push({
          text: sentence,
          score: analysis.comparative,
          provider: result.provider,
        });
      }
    }

    const scores = allSentences.map((s) => s.score);
    const overall = mean(scores);
    const label = scoreLabel(overall);
    const stddev = standardDeviation(scores);
    // Confidence: higher when stddev is low and we have enough samples
    const sampleFactor = Math.min(allSentences.length / 10, 1);
    const varianceFactor = Math.max(0, 1 - stddev);
    const confidence = Math.round(sampleFactor * varianceFactor * 100) / 100;

    let positiveCount = 0;
    let neutralCount = 0;
    let negativeCount = 0;
    for (const s of allSentences) {
      const l = scoreLabel(s.score);
      if (l === 'positive') positiveCount++;
      else if (l === 'negative') negativeCount++;
      else neutralCount++;
    }

    // By provider
    const byProvider: Record<string, number> = {};
    const byProviderLabel: Record<string, 'positive' | 'neutral' | 'negative'> = {};
    const providerGroups = groupBy(allSentences, (s) => s.provider);
    for (const [provider, providerSentences] of providerGroups) {
      const providerScore = mean(providerSentences.map((s) => s.score));
      byProvider[provider] = providerScore;
      byProviderLabel[provider] = scoreLabel(providerScore);
    }

    // By category — use queryId to look up category from query results
    const byCategory: Record<string, number> = {};
    const categoryGroups = groupBy(validResults, (r) => {
      // Extract category from queryId prefix: kw=keyword, ps=persona, cp=competitor, in=intent
      const prefix = r.queryId.split('-')[0];
      const map: Record<string, string> = { kw: 'keyword', ps: 'persona', cp: 'competitor', in: 'intent' };
      return map[prefix] || 'unknown';
    });
    for (const [category, catResults] of categoryGroups) {
      const catScores: number[] = [];
      for (const result of catResults) {
        const sentences = extractSentences(result.response)
          .filter((s) => containsBrand(s, profile.name));
        for (const sentence of sentences) {
          catScores.push(sentiment.analyze(sentence).comparative);
        }
      }
      if (catScores.length > 0) {
        byCategory[category] = mean(catScores);
      }
    }

    // Top positive & negative excerpts
    const sorted = [...allSentences].sort((a, b) => b.score - a.score);
    const topPositive: SentimentExcerpt[] = sorted
      .filter((s) => s.score > 0.05)
      .slice(0, 5)
      .map((s) => ({ text: s.text.slice(0, 200), score: s.score, provider: s.provider }));

    const topNegative: SentimentExcerpt[] = sorted
      .filter((s) => s.score < -0.05)
      .slice(-5)
      .reverse()
      .map((s) => ({ text: s.text.slice(0, 200), score: s.score, provider: s.provider }));

    return {
      overall,
      label,
      confidence,
      byProvider,
      byProviderLabel,
      byCategory,
      positiveCount,
      neutralCount,
      negativeCount,
      totalSentences: allSentences.length,
      topPositive,
      topNegative,
    };
  }
}
